Fix count(*) query over Parquet files

majetideepak opened the pull request:
As reported in https://github.com/facebookexternal/presto_cpp/issues/616, a `select count(*) from t` query over Parquet files fails with an error `failed: !filter_.empty() need a valid filter`.

Fix by removing the check in the `ColumnSelector::getProjection()` method that verifies that `ColumnFilter` is not empty. It is possible that the `ColumnFilter` is empty as in the `select count(*)` case.

majetideepak commented on velox/dwio/common/ColumnSelector.cpp:
```diff
@@ -248,7 +248,6 @@ void ColumnSelector::setReadAll() {
 }
 
 const ColumnFilter& ColumnSelector::getProjection() const {
-  DWIO_ENSURE(!filter_.empty(), "need a valid filter");
```
It is possible to have an empty projection as in `select count(*)` case. Also, this check is present only here. Other uses of `filter_` like in `ColumnSelector::buildSelectedReordered()` do not have this.

majetideepak commented on velox/dwio/parquet/reader/ParquetReader.cpp:
```diff
@@ -161,7 +161,7 @@ uint64_t ParquetRowReader::next(uint64_t /*size*/, velox::VectorPtr& result) {
         &pool_,
         rowType_,
         BufferPtr(nullptr),
-        columns[0]->size(),
```
columns might be empty.

majetideepak commented on velox/dwio/parquet/tests/ParquetTableScanTest.cpp:
```diff
@@ -86,3 +86,25 @@ TEST_F(ParquetTableScanTest, basic) {
   assertQuery(plan, {split}, "SELECT min(a), max(b) FROM tmp");
 }
+
+TEST_F(ParquetTableScanTest, emptyProjection) {
```
I will update this test based on https://github.com/facebookincubator/velox/pull/873/

yingsu00 commented:
LGTM, just waiting for the tests.

mbasmanova commented on velox/dwio/parquet/tests/ParquetTableScanTest.cpp:
```diff
@@ -86,3 +91,25 @@ TEST_F(ParquetTableScanTest, basic) {
+                  .singleAggregation({}, {"count(0)"})
+                  .planNode();
+
+  assertQuery(plan, {split}, "SELECT count(*) FROM tmp");
```
nit: This test doesn't really need to create a table in DuckDB as it can simply assertQuery(plan, {split}, "SELECT 20")

mbasmanova commented on velox/dwio/parquet/tests/ParquetTableScanTest.cpp:
```diff
@@ -86,3 +91,25 @@ TEST_F(ParquetTableScanTest, basic) {
   assertQuery(plan, {split}, "SELECT min(a), max(b) FROM tmp");
 }
+
+TEST_F(ParquetTableScanTest, emptyProjection) {
```
nit: perhaps, "count" is a more descriptive name

mbasmanova reviewed (commented):
@majetideepak Deepak, thank you for fixing this. It would be nice to clarify the problem and the solution in the PR description and update the title to say what this change does, not what was happening before, e.g. "Fix count(*) query over Parquet files"

majetideepak commented:
@mbasmanova thanks for the feedback. I will be more diligent on the PR title and description moving forward. I addressed your comments. Let me know if there are other thoughts. Thanks.

zzhao0 commented:
@majetideepak, does this problem happen for dwrf format as well? is it because of https://github.com/facebookincubator/velox/blob/main/velox/connectors/hive/HiveConnector.cpp#L470?
